src/libappimage/desktop_integration/Thumbnailer.cpp

```diff
@@ -0,0 +1,156 @@
+
+            auto normalIconData = getIconData(resources, appIcon, "128x128");
+            generateNormalSizeThumbnail(canonicalPathMd5, normalIconData);
+
+            auto largeIconData = getIconData(resources, appIcon, "256x256");
+            generateLargeSizeThumbnail(canonicalPathMd5, largeIconData);
+        }
+
+        void Thumbnailer::remove(const core::AppImage& appImage) {
+            std::string canonicalPathMd5 = getCanonicalPathMd5(appImage.getPath());
+            bf::path normalThumbnailPath = getNormalThumbnailPath(canonicalPathMd5);
+            bf::path largeThumbnailPath = getLargeThumbnailPath(canonicalPathMd5);
+
+            bf::remove(normalThumbnailPath);
+            bf::remove(largeThumbnailPath);
+        }
+
+        void Thumbnailer::generateNormalSizeThumbnail(const std::string& canonicalPathMd5,
+                                                      std::vector<char>& normalIconData) const {
+
+            bf::path normalThumbnailPath = getNormalThumbnailPath(canonicalPathMd5);
+            bf::create_directories(normalThumbnailPath.parent_path());
+
+            try {
+                std::vector<uint8_t> iconData{normalIconData.begin(), normalIconData.end()};
+                IconHandle iconHandle(iconData);
+                iconHandle.setSize(128);
+                iconHandle.save(normalThumbnailPath.string(), "png");
+                return;
+            } catch (const IconHandleError&) {}
+
+            // It wasn't possible to generate a thumbnail, therefore the the icon will be written as it's
+            bf::ofstream out(normalThumbnailPath);
+            out.write(normalIconData.data(), normalIconData.size());
+        }
+
+        void Thumbnailer::generateLargeSizeThumbnail(const std::string& canonicalPathMd5,
+                                                     std::vector<char>& largeIconData) const {
+            bf::path largeThumbnailPath = getLargeThumbnailPath(canonicalPathMd5);
+            bf::create_directories(largeThumbnailPath.parent_path());
+
+            try {
+                std::vector<uint8_t> iconData{largeIconData.begin(), largeIconData.end()};
+                IconHandle iconHandle(iconData);
+                iconHandle.setSize(256);
+                iconHandle.save(largeThumbnailPath.string(), "png");
+                return;
+            } catch (const IconHandleError&) {}
+
+            // It wasn't possible to generate a thumbnail, therefore the the icon will be written as it's
+            bf::ofstream out(largeThumbnailPath);
+            out.write(largeIconData.data(), largeIconData.size());
+            out.close();
+        }
+
+
+        std::string Thumbnailer::getCanonicalPathMd5(const std::string& appImagePath) const {
+            auto canonicalAppImagePath = bf::weakly_canonical(appImagePath).string();
+            auto md5Raw = appimage_get_md5(canonicalAppImagePath.c_str());
```

Comment:
Would be nicer when a C++ interface would be used, returning a more safe `std::array`, which would then be converted to a string with proper size information, avoiding buffer overflow attacks (never trust your own code).

Reply:
Done.